### DISCUSSION TITLE
关于接口返回数据封装方式的问题

### DISCUSSION THREAD

**Discussion by Myaem** [zuihou/zuihou-admin-cloud#65] [02.提问]
# 版本信息：
JDK 版本(必填) : 1.8.0
源码版本(必填)：
MySQL(必填)： 5.7.9
Nacos(必填)：

# 问题描述：
您好，在您提供的demo中（图1），我看到所有的controller返回数据Bean，`R`均在接口方法里进行了封装（图2）。
我之前在别地方看到有一种方式是接口只返回原始数据，而数据 Bean `R`在`@RestControllerAdvice`注解的Advice类里进行统一封装（图3）。
请问您觉得这两种方式各自的优缺有哪些，哪一种更适合？期待您的回复，谢谢！

# 相关截图
![](https://s3.ax1x.com/2020/12/24/r2KS4P.png)
![](https://s3.ax1x.com/2020/12/24/r2lK8P.png)
![](https://s3.ax1x.com/2020/12/24/r23NcV.png)

# 支持一下
如果该issue帮助了你, 麻烦给项目点个star, 并给当前issue回复一下 “👍” 或者 “已解决”.

**Comment by zuihou** [comment]
1. 手写R比较简单、灵活， 在代码中有较多判断，并返回不同参数给前端时推荐手写R<Object> 返回。（如代码需要实现以下逻辑，而又不能用抛异常代替时，推荐手写R返回）
```
if (xx){
    return R.fail();
}
```
2. 而`@RestControllerAdvice` 实现全局统一返回则能减少冗余代码，是提高代码质量不错的解决方案。
3. 简单逻辑推荐使用 `@RestControllerAdvice`， 但最好能加一些逻辑（如忽略统一返回的注解），用于控制忽略全局RestControllerAdvice返回，用于兼容某些特殊接口不是按照R的格式返回的。
